task/host-sp-comms/src/bsp/gimlet_bcd.rs

```diff
@@ -0,0 +1,674 @@
+                else { unreachable!(); };
+            *identity = read_one_barcode(dev, &[(*b"BARC", 0)])?.into();
+            Ok(&data)
+        })
+    }
+
+    /// Reads the fan EEPROM barcode values
+    ///
+    /// The fan EEPROM includes nested barcodes:
+    /// - The top-level `BARC`, for the assembly
+    /// - A nested value `SASY`, which contains four more `BARC` values for each
+    ///   individual fan
+    ///
+    /// On success, packs the barcode into `self.tx_buf`; on failure, return an
+    /// error (`DeviceAbsent` if we saw `NoDevice`, or `DeviceFailed` on all
+    /// other errors).
+    fn read_fan_barcodes(
+        &mut self,
+        sequence: u64,
+        name: &[u8],
+        f: fn(userlib::TaskId) -> I2cDevice,
+    ) {
+        let dev = f(I2C.get_task_id());
+        let mut data = InventoryData::FanIdentity {
+            identity: Default::default(),
+            vpd_identity: Default::default(),
+            fans: Default::default(),
+        };
+        self.tx_buf.try_encode_inventory(sequence, name, || {
+            let InventoryData::FanIdentity {
+                identity,
+                vpd_identity,
+                fans: [fan0, fan1, fan2]
+            } = &mut data else { unreachable!(); };
+            *identity = read_one_barcode(dev.clone(), &[(*b"BARC", 0)])?.into();
+            *vpd_identity =
+                read_one_barcode(dev.clone(), &[(*b"SASY", 0), (*b"BARC", 0)])?
+                    .into();
+            *fan0 =
+                read_one_barcode(dev.clone(), &[(*b"SASY", 0), (*b"BARC", 1)])?
+                    .into();
+            *fan1 =
+                read_one_barcode(dev.clone(), &[(*b"SASY", 0), (*b"BARC", 2)])?
+                    .into();
+            *fan2 =
+                read_one_barcode(dev.clone(), &[(*b"SASY", 0), (*b"BARC", 3)])?
+                    .into();
+            Ok(&data)
+        })
+    }
+}
+
+/// Free function to read a nested barcode, translating errors appropriately
+fn read_one_barcode(
+    dev: I2cDevice,
+    path: &[([u8; 4], usize)],
+) -> Result<oxide_barcode::VpdIdentity, InventoryDataResult> {
+    let eeprom = At24Csw080::new(dev.clone());
+    let mut barcode = [0; 32];
+    match drv_oxide_vpd::read_config_nested_from_into(
+        eeprom,
+        path,
+        &mut barcode,
+    ) {
+        Ok(n) => {
+            // extract barcode!
+            let identity = oxide_barcode::VpdIdentity::parse(&barcode[..n])
+                .map_err(|_| InventoryDataResult::DeviceFailed)?;
+            Ok(identity)
+        }
+        Err(
+            VpdError::ErrorOnBegin(err)
+            | VpdError::ErrorOnRead(err)
+            | VpdError::ErrorOnNext(err)
+            | VpdError::InvalidChecksum(err),
+        ) if err
+            == tlvc::TlvcReadError::User(EepromError::I2cError(
+                ResponseCode::NoDevice,
+            )) =>
+        {
+            // TODO: ringbuf logging here?
```

Comment:
"do you want to address this before merging" check

Reply:
Nah, deleted the comment